test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+
+        (uint256 pledgerWeight, uint256 totalWeight) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        DominantJuice.Campaign memory campaign = dominantJuice.getCampaignInfo();
+
+        uint256 pledgerRefundBonus = (pledgerWeight / totalWeight) * campaign.totalRefundBonus;
+
+        if (dominantJuice.getBalance() < pledgerRefundBonus) {
+            vm.expectRevert(abi.encodeWithSelector(DominantJuice.InsufficientFunds.selector));
+        }
+
+        vm.expectEmit(true, true, true, true, address(dominantJuice));
+        emit CycleRefundBonusWithdrawal(pledger1, pledgerRefundBonus);
+
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice.didRedeem(didRedeemData);
+
+        assertEq(true, dominantJuice.getPledgerRefundStatus(pledger1));
+        assertEq(0, dominantJuice.getBalance());
+        assertEq(pledgerRefundBonus, TOTAL_REFUND_BONUS);
+        // Since pledger1 pays through the JB architecture, it doesn't subtract from pledger1's balance.
+        // The unit test shouldn't cover outside state, but this was just an additional sanity assert.
+        assertEq(STARTING_BALANCE + pledgerRefundBonus, pledger1.balance);
+    }
+
+    // Add more didRedeem fuzz tests with multiple redeemers
+
+    // EDGE CASE - contract has less balance than it should. How to mock?
+
+    //////////////////////////////
+    // creatorWithdraw() Tests
+    //////////////////////////////
+
+    function test_creatorWithdraw_revertsForNonOwner() public bonusDeposited {
+        // vm.assume(_notOwner != campaignManager && _notOwner != address(0));
+        successfulCycleHasExpired();
+
+        vm.expectRevert(
+            "AccessControl: account 0x8e24d86be44ab9006bd1277bddc948ecebbfbf6c is missing role 0x5022544358ee0bece556b72ae8983c7f24341bd5b9483ce8a19bff5efbb2de92"
+        );
+        vm.prank(rando);
+        dominantJuice.creatorWithdraw(rando, TOTAL_REFUND_BONUS);
+    }
+
+    function testFuzz_creatorWithdraw_revertsBeforeUnlock(uint256 _seconds) public bonusDeposited {
+        vm.assume(_seconds < cycleExpiryDate + 1209600);
+        vm.warp(_seconds);
+
+        vm.expectRevert("Cycle must be expired and successful, or it must be passed the lock period.");
+        vm.prank(campaignManager);
+        dominantJuice.creatorWithdraw(campaignManager, TOTAL_REFUND_BONUS);
+    }
+
+    function test_creatorWithdraw_revertsIfGoalNotMet() public bonusDeposited {
+        failedCycleHasExpired();
+
+        vm.expectRevert("Cycle must be expired and successful, or it must be passed the lock period.");
+        vm.prank(campaignManager);
+        dominantJuice.creatorWithdraw(campaignManager, TOTAL_REFUND_BONUS);
+    }
+
+    function testFuzz_creatorWithdraw_revertsOnOverdraw(uint256 _amount) public bonusDeposited {
+        successfulCycleHasExpired();
+
+        vm.assume(_amount > dominantJuice.getBalance());
```

Comment:
you could get a _lot_ of retests since the balance is in ether. Maybe bound instead?

Reply:
I can do it with bound instead.... But what do you mean by retests? Do you mean false passing tests?

Reply:
sorry, if you use assume, my understanding is that the forge test vm fails with a special revert message that then causes the test to be rerun with a different value, so it's like hunting and pecking for usable values. If the value space is large, that can take a long time, randomly

Reply:
Ah I see. The fuzz tests are almost as fast as regular tests. I haven't really come across any major lags (aside from my 5-year-old computer!). `vm.assume` use here is basically acting like a one-sided `bound`, so it'll just try 250 random values greater than `dominantJuice.getBalance()` and will stop the test at the first value that fails the test.

Reply:
cool